Cross Validation

wangcj05 opened the pull request:
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #327 

##### What are the significant changes in functionality due to this change request?
Add a new post processor to perform cross validation.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

wangcj05 commented:
@alfoa This is the initial implementation, and I believe a lot of things need to be changed. If you have time, please take a look and let me know if you want to change something.

wangcj05 commented:
@alfoa I will start to clean my branch, add more tests, and documentations.

wangcj05 commented:
@alfoa Do you have time to take a look at this PR?

alfoa commented:
ok

aalfonsi commented on framework/CrossValidations/Factory.py:
```diff
@@ -0,0 +1,80 @@
+## [ Add new class here ]
+################################################################################
+## Alternatively, to fully automate this file:
+# from OutStreamManagers import *
```
we can delete this (or better link the description to this class)

wangcj05 replied:
Ok, I will delete this.

aalfonsi commented on tests/framework/PostProcessors/CrossValidations/gold/simpleCrossValidation/dumpMC.xml:
```diff
@@ -0,0 +1,370 @@
+<data name="dumpMC" type="Pointset"><input>x2,x1</input><output>ans2,ans</output><inputFilename>dumpMC.csv</inputFilename><metadata><SamplerType>array([u'MonteCarlo', u'MonteCarlo', u'MonteCarlo', u'MonteCarlo',
```
do we really need this outputfile?

wangcj05 replied:
We do not need this, I will delete it.

aalfonsi commented:
Everything looks good except a couple of comments... 

Thanks

wangcj05 commented:
Hi @alfoa @joshua-cogliati-inl,
When I tried to add a map between the parameters, I found the 'cross_validation' module is still available in scikit learn version 0.18 and 0.19, but they are deprecated and will be removed in 0.20.

I think for now it is ok to just use the module 'cross_validation', and moved to 'model_selection' when our minimum requirement of scikit learn is 0.20. The other reason for this is that the call for this methods is changed in module 'model_selection'. Please let me know what do you think.

joshua-cogliati-inl commented:
Looking at the differences, it looks like in the 0.18 model_selection the n parameter is automatically calculated from the data.  I think that could be easily calculated from the data if we are in 0.17.  Are there other changes that we can insulate the user from?  Basically, if cross_validation is 0.17 - 0.19, I am fine with using it, but if switching will cause RAVEN input file changes, I would like to insulate the user as much as possible.

wangcj05 commented:
Yes, the methods inside module 'cross_validation' are consistence in versions 0.17~0.19. We do not need to change the input file when switching. But from 0.18, the way to call the cross validation methods inside 'model_selection' module and 'cross_validation' module is different. That's why I suggest to use 'cross_validation' module.

alfoa commented:
Agree with Conjiang

joshua-cogliati-inl commented:
If we do not need to change the input when we switch to model_selection then I have no objection to using cross_validation now. (Agree with Congjian )

wangcj05 commented:
@joshua-cogliati-inl Sorry if I did not express myself well. We do not need to change the input across the versions when using 'cross_validation' module, but the problem is that this module will be removed in 0.20. There are some input differences between 'cross_validation' and 'model_selection' which will lead the change of raven input file. Besides, some method names are also changed between 'cross_validation' and 'model_selection' modules, for example method 'LabelKFold' changed to 'GroupKFold'. Moreover, if we use 'cross_validation', we can just call the method like 'KFold(n_folds=3)' to get the train and test indices. For the 'model_selection' module, we need to call method like 'kf=KFold(n_splits=3)', then 'kf.split(inputData)' to generate the indices. Please let me know what do you think.

joshua-cogliati-inl commented:
@wangcj05 Tell me if I am wrong. So KFold would be able to work both with model_selection and cross_validation since the parameters are functionally the same, and the user input would not need to change.  However as I understand it you are saying that some of the other methods have different enough inputs that we would not easily be able to do a conversion like that.  Are any of the other methods needed now (or before we switch to sklearn 0.18?)?  If not, maybe we can just leave them undocumented, and after we switch to 0.18 we can document them then.

joshua-cogliati-inl commented:
Also, I think this pull request can go in even if we will have to change the RAVEN inputs when we start supporting 0.20, but I would rather not have to change the input if not too much effort.

wangcj05 commented:
Ok, I made some changes to make sure the input will be consistence. Based on my understanding, when we switch to use 'model_selection', the input can be the same, we only need to make the parameter 'n' optional instead of required.

wangcj05 commented:
I think this PR is ready to go in. @alfoa @joshua-cogliati-inl Please let me know if you have any questions.

aalfonsi commented:
----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code. ***ok***
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).  ***new feature. Documented in user manual***
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.  ***ok***
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.  ***ok***
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.  ***tests added***
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.  ***n/a***
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.  ***#327***
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?  ***n/a***

aalfonsi commented:
Checklist passed...

Merging...
